src/ray/object_manager/object_manager.cc

```diff
@@ -196,19 +207,48 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
   return ray::Status::OK();
 }
 
-ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id,
-                                int retry) {
+void ObjectManager::HandlePushTaskTimeout(const ObjectID &object_id,
+                                          const ClientID &client_id) {
+  RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
+                 << " after waiting for " << config_.push_timeout_ms << " ms.";
+  auto iter = unfulfilled_push_tasks_.find(object_id);
+  if (iter != unfulfilled_push_tasks_.end()) {
+    iter->second.erase(client_id);
+    if (iter->second.size() == 0) {
+      unfulfilled_push_tasks_.erase(iter);
+    }
+  }
+}
+
+ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
   if (local_objects_.count(object_id) == 0) {
-    if (retry < 0) {
-      retry = config_.max_push_retries;
-    } else if (retry == 0) {
-      RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
-                     << " after retrying " << config_.max_push_retries << " times.";
-      return ray::Status::OK();
+    // Avoid setting duplicated timer for the same object and client pair.
+    auto &clients = unfulfilled_push_tasks_[object_id];
+    if (clients.count(client_id) == 0) {
+      // If config_.push_timeout_ms < 0, we give an empty timer
+      // and the task will be kept infinitely.
+      auto timer = std::shared_ptr<boost::asio::deadline_timer>();
+      if (config_.push_timeout_ms == 0) {
+        // The Push request fails directly when config_.push_timeout_ms == 0.
+        RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
+                       << " due to direct failure setting. ";
+        return ray::Status::OK();
+      } else if (config_.push_timeout_ms > 0) {
+        // Put the task into a queue and wait for the notification of Object added.
+        timer.reset(new boost::asio::deadline_timer(*main_service_));
+        auto clean_push_period = boost::posix_time::milliseconds(config_.push_timeout_ms);
+        timer->expires_from_now(clean_push_period);
+        timer->async_wait(
+            [this, object_id, client_id](const boost::system::error_code &error) {
+              // Timer killing will receive the boost::asio::error::operation_aborted,
+              // we only handle the timeout event.
+              if (!error) {
+                HandlePushTaskTimeout(object_id, client_id);
+              }
+            });
+      }
+      clients.emplace(client_id, timer);
```

Comment:
When `config_.push_timeout_ms == 0`, there is no reason to add a null timer to `unfulfilled_push_tasks_`, right?

Reply:
Yes, when `config_.push_timeout_ms == 0` we don't need to add this record. However,  when `config_.push_timeout_ms < 0`, we need to add this record, because -1 means waiting infinitely.